### PR TITLE
NSG support MetricType IP

### LINKED ISSUE #533: NSG build failed with cpu-version
**Describe the bug**
```
2019-11-26 14:00:56,054 | ERROR | default | [ENGINE] Table file doesn't exist: /opt/milvus/db/tables/annglove_1m_1024_200_ip/1191026/1574747755039365000 in path: /opt/milvus/db for table: annglove_1m_1024_200_ip
2019-11-26 14:00:56,225 | ERROR | default | [WRAPPER] Error in virtual knowhere::IndexModelPtr knowhere::NSG::Train(const DatasetPtr&, const Config&) at IndexNSG.cpp:119: NSG not support this kind of metric type
2019-11-26 14:00:56,225 | ERROR | default | [ENGINE] BuildIndex encounter exception: Error in virtual knowhere::IndexModelPtr knowhere::NSG::Train(const DatasetPtr&, const Config&) at IndexNSG.cpp:119: NSG not support this kind of metric type
2019-11-26 14:00:56,227 | ERROR | default | [ENGINE] Failed to build index, index file is too large or gpu memory is not enough
2019-11-26 14:00:56,227 | ERROR | default | [ENGINE] Building index failed: Error: BuildIndex encounter exception: Error in virtual knowhere::IndexModelPtr knowhere::NSG::Train(const DatasetPtr&, const Config&) at IndexNSG.cpp:119: NSG not support this kind of metric type
```

**Steps/Code to reproduce behavior**
dataset: glove-200-ip

**Expected behavior**

**Environment details**
0.6.0-CPU-version

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by tinkerlin**
Fixes #533

### CODE REVIEW

**Review by youny626 — APPROVED** [review]

File: core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    }
+#endif
+#endif
+#endif
+
+    return result;
+}
+
+float
+DistanceIP::Compare(const float* a, const float* b, unsigned size) const {
```
Comment (cydrain) [inline core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.cpp]: why unsigned ?
can use int32_t or int64_t ?

File: core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+namespace knowhere {
+namespace algo {
+
+struct Distance {
```
Comment (cydrain) [inline core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.h]: why struct not class ?

File: core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.cpp (hunk 1 above)
Comment (tinkerlin) [reply core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.cpp]: No

File: core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.h (hunk 2 above)
Comment (tinkerlin) [reply core/src/index/knowhere/knowhere/index/vector_index/nsg/Distance.h]: should be struct

**Review by cydrain — APPROVED** [review]

**Pull request merged**